doc/codeExecution.md

```diff
@@ -0,0 +1,115 @@
+* The user must mount the Docker socket into PrairieLearn's container, similarly to what they must do when running external grading locally.
+* The user must create a "scratch" directory on the host, specify it with the `HOSTFILES_DIR` environment variable, and mount the directory to `/hostfiles` in the PrairieLearn container. PrairieLearn will use this directory to communicate its internal files, including element implementations and the Python zygote, to the host so that they can be mounted into executor containers.
+
+PrairieLearn will then maintain a pool of workers, one per course, with the corresponding host course directory mounted into each container. It will also begin watching the `elements/`, `python/`, and `exampleCourse/` directories and copying their files to `HOSTFILES_DIR` so that they can be accessible to the executor containers. Most of the time, this copy will only take place one time when PrairieLearn boots up, but the continuous copying can be useful if you're running the PrairieLearn container with your own copy of the PrairieLearn code mounted to `/PrairieLearn`. When you edit files in your own copy, they'll magically be copied back to the right place on the host and things will Just Work(tm).
+
+That was a lot of words; let's look at some concrete configs and Docker commands.
+
+Let's assume you have some PrairieLearn course on your machine at `/Users/nathan/git/pl-cs225`. Here's your corresponding config file.
+
+```json
+{
+  "courseDirs": [
+    "/course"
+  ],
+  "courseDirsHost": {
+    "/course": "/Users/nathan/git/pl-cs225"
+  },
+  "workersExecutionMode": "container"
+}
+```
+
+Now, we're ready to run PrairieLearn. Say you saved the above config file in `/Users/nathan/config.json`. You can now run the following Docker command:
+
+```sh
+docker run --rm -it -p 3000:3000 -v /Users/nathan/git/pl-cs225:/course -v /Users/nathan/config.json:/PrairieLearn/config.json -v /Users/nathan/.plhostfiles:/hostfiles -e HOST_JOBS_DIR=/Users/nathan/.pl_ag_jobs -e HOSTFILES_DIR=/Users/nathan/.plhostfiles -v /var/run/docker.sock:/var/run/docker.sock prairielearn/prairielearn
+```
+
+## Code execution in practice
+
+So far, this discussion has been pretty abstract. But what about all the actual code that underpins this stuff? Fear not, dear reader, we haven't forgotten about that!
+
+### Code callers
+
+A *code caller* serves as an abstraction on top of the different execution modes above and hides the implementation details of exactly how code is executed. There are currently two different types of code callers, referred to here by the filenames of their implementations.
+
+* `lib/code-caller-docker` handles executing code inside of Docker containers, as required by the `container` and `native` execution modes.
+* `lib/code-caller-python` handles executing Python processes directly, as required by the `internal` execution mode.
```

Comment:
IIRC functionality is the same, the one in the executor image just has types added. The cleanest way to share things would likely be to publish this as a package so that both PL and the executor image could consume it. Otherwise, we're either copying a file from another part of the repo into the executor at build time, or we're having PL itself reach into the `images/executor/src` directory, which I don't very much like. This code is largely stable though, so I'm not too focused on duplication at the moment.

Reply:
Sounds reasonable to me!